File: src/folkman/solver/folkman.py
# Folkman.py
# Author: Christopher Wood, caw4567@rit.edu

# Don't reinvent the wheel...
from networkx import nx

def checkPresence(F, n, edgeColor, edgeColorMap, color):
	''' Check to see if a K_n of color 'color' is in F.

	Note: edgeColor maps edge indices to colors, and edgeColorMap maps edges 
	to edge indices, use as needed.
	'''
	if (n == 3): # Search for triangles with a stupid simple for loop...
		edgeIndex = 0
		for edgeIndex in range(0, len(edgeColor)):
			edge = F.edges()[edgeIndex]
			if (edgeColor[edgeIndex] == color): # Only look at the specific color
				for vertex in F.nodes():
					# Format the edge for linst the index check
					edge1 = ()
					if (edge[0] < vertex):
						edge1 = (edge[0], vertex)
					else:
						edge1 = (vertex, edge[0])
					edge2 = ()
					if (edge[1] < vertex):
						edge2 = (edge[1], vertex)
					else:
						edge2 = (vertex, edge[1])

					# Check to see if we have a triangle of the same color
					if (edge1 in F.edges() and edge2 in F.edges()):
						if (edgeColor[edgeColorMap[edge1]] == color and edgeColor[edgeColorMap[edge2]] == color):
							return True
		return False 
	elif (n == 4):
		print("TODO: 2/7/13")
	else:
		raise Exception("Cases other than n = 3,4 are not yet implemented.")
	return False

def walkColors(F, m, n, edgeColor, edgeColorMap, edgeIndex, maxColor, colorSet):
	''' Walk over all possible colorings for the graph F.

	Note: edgeColor maps edge indices to colors, and edgeColorMap maps edges to edge 
	indices, use as needed. Also, maxColor is the maximum number of colors to check, 
	colorSet is the result of each edge coloring, and m/n are the order of the 
	complete graphs to check.
	'''
	if (edgeIndex == (len(edgeColor) - 1)): # base case - last edge to color
		for i in range(0, maxColor):
			edgeColor[edgeIndex] = i
			induced = False

			# We've completed the color assignment... now check to see if there's a K_m in color1/2 or K_n in color1/2
			for color in range(0, maxColor):
				if (checkPresence(F, m, edgeColor, edgeColorMap, color)):
					induced = True
				elif checkPresence(F, n, edgeColor, edgeColorMap, color):
					induced = True

			# Dump in the result
			colorSet.append(induced)
	else: # keep coloring more edges recursively
		foundSubgraph = False
		for i in range(0, maxColor):
			edgeColor[edgeIndex] = i
			result = walkColors(F, m, n, edgeColor, edgeColorMap, edgeIndex + 1, maxColor, colorSet)
			if (result):
				foundSubgraph = True
		return foundSubgraph

def folkman(F, m, n):
	''' Decide whether or not F -> (K_m,K_n).
	'''
	# Initialize the edge color sets
	edgeColorMap = {}
	edgeColor = []
	index = 0
	for edge in F.edges():
		edgeColorMap[edge] = index
		edgeColor.append(0) # start with color 0
		index = index + 1

	# Walk the colors and do the check...
	colorSet = []
	walkColors(F, m, n, edgeColor, edgeColorMap, 0, 2, colorSet) # there's only two colors here...
	for i in colorSet:
		if (i == False):
			return False
	return True

def main():

	# TODO: read in all three params from command line
	# TODO: first must be a graph (adj matrix), second can be graphs or 

	# The two test cases shown in the presentation...
	F = nx.complete_graph(5)
	print("Checking K_5 -> (3,3)")
	print(folkman(F, 3, 3))

	F = nx.complete_graph(6)
	print("Checking K_6 -> (3,3)")
	print(folkman(F, 3, 3))

if __name__ == "__main__":
	main()

File: src/folkman/solver/folkman.cpp
// Folkman.cpp
// Author: Christopher Wood, caw4567@rit.edu

#include <stdio.h>
#include <vector>

// Storage container for an edge (used in the edge map)
typedef struct 
{
	int x;
	int y;
} edge;

// For convenience...
using namespace std;

// Store the results - NO LONGER USED BECAUSE OF MAXIMUM SIZE
vector<bool> results;

// Display the adjacency matrix for the graph F
void display(int order, int** F)
{
	for (int i = 0; i < order; i++)
	{
		for (int j = 0; j < order; j++)
		{
			printf("%d ", F[i][j]);
		}
		printf("\n");
	}
}

// Check to see if K_n is induced by the edges colored in F by 'color'
bool isInduced(int order, int n, int edgeCount, int** F, int** colorState, edge* edgeMap, int color)
{
	bool induced = false;

	if (n == 3) // special case for triangles
	{
		for (int edgeIndex = 0; edgeIndex < edgeCount; edgeIndex++) // search over each  edge
		{
			int x = edgeMap[edgeIndex].x;
			int y = edgeMap[edgeIndex].y;
			if (colorState[x][y] == color)
			{
				for (int v = 0; v < order && (x != v) && (y != v); v++) // check each of the remaining vertices
				{
					if (F[x][v] == 1 && F[y][v] == 1 && colorState[x][v] == color && colorState[y][v] == color) // both edges exist
					{
						return true;
					}
				}
			}
		}
	}
	else if (n == 4)
	{
		for (int edgeIndex = 0; edgeIndex < edgeCount; edgeIndex++) // search over each  edge
		{
			int x = edgeMap[edgeIndex].x;
			int y = edgeMap[edgeIndex].y;
			if (colorState[x][y] == color)
			{
				for (int v = 0; v < order && (x != v) && (y != v); v++) // check each of the remaining vertices
				{
					if (F[x][v] == 1 && F[y][v] == 1 && colorState[x][v] == color && colorState[y][v] == color) // both edges exist
					{
						// TODO: continue check for fourth vertex thats in the set...
						for (int v2 = 0; v2 < order && (v2 != v) && (v2 != x) && (v2 != y); v2++)
						{
							if (F[v2][x] == 1 && F[v2][y] && F[v2][v] && 
								colorState[v2][x] == color && colorState[v2][y] == color && colorState[v2][v] == color)
							{
								return true;
							}
						}
					}
				}
			}
		}
	}
	else 
	{
		printf("Error: only n = 3,4 clique sizes are supported.\n");
	}

	return induced;
}

// Walk over all possible colors in the graph F, invoking the k-clique check algorithm for each color
bool walkColors(int order, int color, int** F, int* colors, int** colorState, int edgeCount, edge* edgeMap, int edgeIndex)
{
	// Base case
	if (edgeIndex == edgeCount - 1)
	{
		int x = edgeMap[edgeIndex].x;
		int y = edgeMap[edgeIndex].y;
		for (int c = 0; c < color; c++)
		{
			colorState[x][y] = c;
			colorState[y][x] = c;
			bool induced = false;
			for (int ci = 0; ci < color; ci++)
			{
				if (isInduced(order, colors[ci], edgeCount, F, colorState, edgeMap, ci))
				{
					induced = true;
				}
			} 
			// TODO: Append results to the vector 
			if (induced == false)
			{
				return induced;
			}
		}
		return true;
	}
	else 
	{
		bool induced = false;
		int x = edgeMap[edgeIndex].x;
		int y = edgeMap[edgeIndex].y;
		for (int c = 0; c < color; c++)
		{
			colorState[x][y] = c;
			colorState[y][x] = c;
			if (walkColors(order, color, F, colors, colorState, edgeCount, edgeMap, edgeIndex + 1))
			{
				induced = true;
			} 
			else {
				return false;
			}
		}
		return induced;
	}
}

// Decide if F -> (s1, s2, ..., sk)^e
bool folkman(int order, int color, int** F, int* colors)
{
	// Allocate space for the color state -  we only use the upper diagnoal of the matrix
	int** colorState = new int*[order];
	for (int i = 0; i < order; i++) 
	{
		colorState[i] = new int[order];
	}

	// Copy over the required portions and set all edge colors to 0
	int edgeCount = 0;
	for (int i = 0; i < order; i++) 
	{
		for (int j = 0; j < order && i >= j; j++) // only look at the upper diagonal
		{
			if (F[i][j] == 1) 
			{
				edgeCount++;
				colorState[i][j] = 0;
				colorState[j][i] = 0;
			}
			else 
			{
				colorState[i][j] = -1; // this edge doesn't exist, so its color is -1
				colorState[j][i] = -1; // Ditto
			}
		}
	}

	// Allocate and initialize the edge map
	int edgeIndex = 0;
	edge* edgeMap = new edge[edgeCount];
	for (int i = 0; i < order; i++) 
	{
		for (int j = 0; j < order; j++) // only look at the upper diagonal
		{
			if (F[i][j] == 1 && j >= i)
			{
				edge e;
				e.x = i;
				e.y = j;
				edgeMap[edgeIndex++] = e;
			}
		}
	}

	printf("edge count contents:\n");
	for (int i = 0; i < edgeCount; i++)
	{
		printf("%d,%d\n", edgeMap[i].x, edgeMap[i].y);
	}

	// Now, walk the colors... but to do it sequentially or recursively... that is the question.
	// Python implementation does it recursively
	bool result = walkColors(order, color, F, colors, colorState, edgeCount, edgeMap, 0);

	/*bool result = true;
	for (int i = 0; i < results.size(); i++)
	{
		if (results[i] == false) 
		{
			result = false;
			break;
		}
	}*/

	return result;
}

// Define the order of the first graph
#define FSIZE 9

int main(int argc, char** argv)
{
	// K_FSIZE for testing color permutation algorithm
	int** F = new int*[FSIZE];
	for (int i = 0; i < FSIZE; i++)
		F[i] = new int[FSIZE];
	for (int i = 0; i < FSIZE; i++)
	{
		for (int j = 0; j < FSIZE; j++)
		{
			if (i == j) F[i][j] = 0;
			else F[i][j] = 1;
		}
	}

	int* colors = new int[2];
	colors[0] = 4;
	colors[1] = 3;

	// Run the checks now...
	printf("K_%d -> (%d, %d)?\n", FSIZE, colors[0], colors[1]);
	bool result = folkman(FSIZE, 2, F, colors);
	if (result) printf("Yes.\n");
	else printf("No.\n");

	return 0;
}
